### DISCUSSION TITLE
Drop the protocol selection in the Installation Source spoke

### DISCUSSION THREAD

**Discussion by poncovka** [rhinstaller/anaconda#4257] [Ideas]
## Introduction

We are planning an extensive clean-up of the code for the Installation Source spoke to prepare the code for the payload modularization. That creates a good opportunity for revisiting the design of the user interface as well. We would like to address some of the long-term issues and propose an alternative. It shouldn't be anything drastic, just a small reorganization of the current components and options.

At this moment, the spoke looks (for example) like this:
![current_state](https://user-images.githubusercontent.com/7396092/183667386-2779469c-f1b5-4a40-a831-7fe068d8ddad.png)

The installer supports several options, that doesn't have to be always available:
* "Auto-detected installation media" is visible only if an installation media is detected.
* "Installation media via SE/HMC" is visible only on s390x with SE/HMC support.
* "ISO file" is visible only if a mountable device is found.
* "Red Hat CDN" is visible only on RHEL.

#### What are the issues?

1. Some of the combinations of options don't make sense. The installer ignores anything irrelevant, so they don't break anything, but it is just confusing. See the "Closest mirror" option.
2. Keeping protocols out of URLs doesn't work very well. When users specify the URL with a protocol, the installer removes the protocol from their input and sometimes shows weird warnings about non-matching protocols even when they match.
3. The base source and additional sources sometimes use different components for the same options. For example, both of them could use the proxy configuration dialog.

#### What would we like to achieve?

* Don't remove protocols from URLs.
* Don't keep protocols outside of URLs.
* Don't show options irrelevant to the selected type.

## Suggestions for the "On the network" option
We have some suggestions for the "On the network" option of the main installation source.

#### Current solution
When you select the HTTP source, it looks like this. HTTPS looks the same.
![current_url](https://user-images.githubusercontent.com/7396092/183669940-db618a8d-683f-4fbc-87c4-312debcb7f9c.png)

When you select the FTP source, it looks like this. The URL type is "repository URL" by default.
![current_ftp](https://user-images.githubusercontent.com/7396092/183669977-f488756a-452a-4cbf-a228-7059f38472d5.png)

#### New solution
The drop-down menu with protocols could be replaced with a drop-down menu with actions. Choose "Install from a specified URL" to see all configuration options related to the URL source. Users could specify an URL of the source, an URL type and a proxy. The "Configure proxy" button would open a dialog for the proxy configuration.
![solution2_url](https://user-images.githubusercontent.com/7396092/183670061-27fe12e3-dd7f-467e-a5c0-47ab8013e4ae.png)

#### Current solution
When you select the NFS source, it looks like this. The URL has to be specified as `<server>:<path>`. The NFS mount options are optional.
![current_nfs](https://user-images.githubusercontent.com/7396092/183670110-ca654fde-e9a7-4389-882c-82433293a32e.png)

#### New solution
Choose "Install from a NFS server" to see all configuration options related to the NFS source. Users could specify a NFS server, a path on the server and optionally NFS mount options.
![solution2_nfs](https://user-images.githubusercontent.com/7396092/183670183-2a34560d-5393-4c56-a8bc-71281301f7e1.png)

#### Current solution
When you select the closest mirror, it looks like this. The closest mirror is the only option that supports installation of updates, so the Updates section is included on this screenshot, but it is visible always. The proxy configuration and the URL type are irrelevant to the closest mirror and ignored by the installer.
![current_closest_mirror](https://user-images.githubusercontent.com/7396092/183670398-e3398afa-712d-40b9-83a6-bb007ad0a394.png)

### New solution
Choose "Install the latest available software updates from the closest mirror" to select the closest mirror with enabled updates.
![solution2_mirror_updates](https://user-images.githubusercontent.com/7396092/183670446-39410ae6-3a5f-4fab-a203-ad46f9143277.png)

Choose "Install the default software versions from the closest mirror" to select the closest mirror with disabled updates.
![solution2_mirror_no_updates](https://user-images.githubusercontent.com/7396092/183670476-a3aff247-cba1-4a37-a798-47c302b62264.png)

The "Updates" section would be removed from the screen.

## Suggestions for the "Additional repositories" section
We have some suggestions for the "Additional repositories" section.

#### Current solution
When you select the HTTP source, it looks like this. HTTPS looks the same.
![current_additional_http](https://user-images.githubusercontent.com/7396092/183670569-c321410e-4bf9-43ee-b4db-7041a04a3397.png)

When you select the FTP source, it looks like this. The URL type is "repository URL" by default.
![current_additional_ftp](https://user-images.githubusercontent.com/7396092/183670621-3cb6fed4-ff23-457a-a228-a5c3d2b872c1.png)

#### New solution
The drop-down menu with protocols could be replaced with a drop-down menu with Source actions. Choose "Install additional software from a specified URL" to see all configuration options related to the URL source in the same way as in the "On the network option".
![solution2_add_url](https://user-images.githubusercontent.com/7396092/183670716-4deb390e-517b-4a51-9c0c-f67aa97e4c80.png)

#### Current solution
When you select the NFS source, it looks like this. The URL has to be specified as `<server>:<path>` or `<options>:<server>:<path>` if you want to specify mount options.

![current_additional_nfs](https://user-images.githubusercontent.com/7396092/183670800-9f0ef6e7-ff8f-460f-85db-98a8d062e95c.png)

#### New solution
Choose "Install additional software from a NFS server" to see all configuration options related to the NFS source in the same way as in the "On the network option".

![solution2_add_nfs](https://user-images.githubusercontent.com/7396092/183670843-6322b869-e42b-472c-88d4-1c10798912f1.png)

### Other locations
In some cases, the installer shows uneditable details of other types of sources.

#### Current solution
We are trying to modify the drop-down menu with protocols and we are not always successful.

#### New solution
The installer will choose "Install additional software from another source" to show details about another type of source. These details won't be editable.

![solution2_add_other](https://user-images.githubusercontent.com/7396092/183670907-5ad7b402-166c-4e5c-8be7-0ebf2c70e6b7.png)

## Summary
With the changes suggested above (and some additional changes in the "ISO file" option), the spoke could look like this:

![solution2_summary_mirror](https://user-images.githubusercontent.com/7396092/183669165-fbd40325-ac4a-446d-a09c-d884350985fe.png)

Or like this when the installation source is switched to URL:

![solution2_summary](https://user-images.githubusercontent.com/7396092/183669187-99bcbcb4-a7a6-4ba0-8ec8-0dcd4587afca.png)

**Comment by poncovka** [comment]
We would also like to rename the "ISO file" option to something like "Device with an ISO file". Using just "ISO file" is confusing, because users can select only an ISO file on a mountable device. They cannot specify for example an URL. Also the device description can be shown on one line if we use an ellipsized label instead of a hard-coded new line. It looks much better.

![main_source_iso_file](https://user-images.githubusercontent.com/7396092/183683187-7b0d68ca-da57-4652-9903-d2bac7515c2d.png)

**Comment by poncovka** [comment]
It might a good idea to hide the "Additional repositories" section by default in an expander.
![expander_1](https://user-images.githubusercontent.com/7396092/183684929-b5fffa46-55c2-44d0-b820-a9f46d64c8c3.png)

**Comment by poncovka** [comment]
@mairin Could you look at this proposal, please? What do you think?

**Comment by jstodola** [comment]
What about unifying the design for the base repository and additional repositories? I mean, place the configuration fields for the base repository vertically instead of horizontally next to each other. There is some space on the screen and it would allow more space for the URL or NFS path.

Also, I'm not sure about the "Install from a specified URL" label. If I were installing from a NFS server, I would try to provide `nfs://<server>:<path>` as the URL. So, what about "Install from an http, https or ftp source"?

Otherwise it looks like a nice improvement.

**Reply by poncovka** [reply]
> What about unifying the design for the base repository and additional repositories? I mean, place the configuration fields for the base repository vertically instead of horizontally next to each other. There is some space on the screen and it would allow more space for the URL or NFS path.

I am not sure if there is enough vertical space when I consider all supported installation source options.

> Also, I'm not sure about the "Install from a specified URL" label. If I were installing from a NFS server, I would try to provide `nfs://<server>:<path>` as the URL. So, what about "Install from an http, https or ftp source"?
> 

I like that. Maybe it could be `Install from an HTTP, HTTPS or FTP server` for consistency with NFS?

**Reply by jstodola** [reply]
I see. Sure.

**Comment by mairin** [comment]
Sorry @poncovka I missed the GitHub notification on this one - in case my review is still helpful:

I like the solutions you came up with here generally, I think they are good simplifications.

### General ideas

To make this screen a bit cleaner / less cluttered, would it be possible to only show options that depend on a dropdown selection when that item is selected? For example, some options apply to https that don't apply to ftp. 

Here's a visual to show what I mean: by default in the "On the network" dropdown, only show that dropdown with a default "please choose option" item selected by default, and fade in / otherwise display the other form fields that apply to the selected drop down option? E.g., something like this:

![Screenshot from 2022-08-18 22-51-31](https://user-images.githubusercontent.com/799683/185532503-2b1de2d8-9508-46a8-93de-fbf244c4c3d3.png)

I also see some areas where the size / alignment / padding of elements is just a bit off and a few visual tweaks might give it overall a nicer appearance - but I know these can be kind of tricky to do in GTK. Let me know if you are interested and I can mock something up to show what I mean.

### Autodetected media

I would look at maybe adjusting the alignment of the "auto detected media section", try this maybe, it might make things feel a bit lighter:
![Screenshot from 2022-08-18 22-38-54](https://user-images.githubusercontent.com/799683/185531087-1b38c309-8d10-42f6-a3cd-61871353d8f3.png)

### For on the network source configuration

- The closest mirror solution is particularly elegant, I really like how you've simplified it here.
- For ftp - it looks like **proxy** and **type** aren't options for ftp in the current version. Do those widgets grey out if you type out a URL that is ftp:// ?
- I love the not required default text for the mount options in the NFS solution!

### Additional repo source configuration

I like your idea of hiding the additional repos configuration area under a disclosure widget - that would make things cleaner and less intimidating to a lot of users who won't use these options but still keep them easily at hand for those who need them.

Another thing to think about - if you applied the idea under "General ideas" at the top of this comment to the additional repo configuration section.... it would make it a lot less cluttered without hiding it below a disclosure triangle. I would still think some kind of horizontal seperator between the main source section and additional sources section would be a good idea, though.


I hope this helps :) Happy to chat more on this if it makes sense.

**Reply by poncovka** [reply]
Hi @mairin! Thank you for your reply! I haven't started to work on these changes yet, so any feedback is appreciated.

> To make this screen a bit cleaner / less cluttered, would it be possible to only show options that depend on a dropdown selection when that item is selected? 

Yes, that's the idea. Anything irrelevant shouldn't be visible. I didn't explain it in the proposal, but basically, what you see on the "New solution" screenshots is what you get. It is the same of the additional repositories. I will add the `Choose network source ...` option, it looks like a nice improvement.

> I also see some areas where the size / alignment / padding of elements is just a bit off and a few visual tweaks might give it overall a nicer appearance - but I know these can be kind of tricky to do in GTK. Let me know if you are interested and I can mock something up to show what I mean.

Yes, I am interested :-)

> I would look at maybe adjusting the alignment of the "auto detected media section".

We can try to address that in a separate issue. At first glance, the suggested solution might be tricky to implement. Every "line" of the installation source section is a separate item of a container, even the ones without a radio button. I don't know how much logic is tied to that arrangement. Hopefully, none.

> For ftp - it looks like proxy and type aren't options for ftp in the current version. Do those widgets grey out if you type out a URL that is ftp://?

The proxy configuration should be available. We support that combination, so I don't know why it is disabled in the current UI. The URL type should be set to `Repository URL` and greyed out based on the typed URL.

**Reply by mairin** [reply]
Hi @poncovka - I filed this poorly on my task list and am just finding it again. Let me know if the padding / appearance tweak ideas would still be helpful and I can schedule them in, but I realize I am following up very late now so they might be less useful.

**Reply by poncovka** [reply]
Hi @mairin! No problem. I have just started to work on these changes, so it might be easier to ping you on the draft pull request and do the tweaks based on the actual implementation. Whatever works for you :-)

**Reply by poncovka** [reply]
The changes for additional repositories are implemented at https://github.com/rhinstaller/anaconda/pull/4498. The changes for the installation source will be implemented separately in another pull request.
